src/model/models.py

```diff
@@ -157,39 +207,70 @@ def __init__(self, config: dict, use_quantum: bool = False):
-                     self.backbone = TCGABackbone(input_dim=768, latent_dim=latent_dim)
+                    self.backbone = TCGABackbone(input_dim=768, latent_dim=latent_dim)
                 else:
-                     self.backbone = PCAMBackbone(latent_dim=latent_dim, filters=model_cfg['pcam_filters'])
-
+                    self.backbone = PCAMBackbone(latent_dim=latent_dim, filters=model_cfg['pcam_filters'])
                 self.head = QuantumHeadAngle(
                     n_qubits=n_qubits,
                     num_classes=num_classes,
                     n_layers=model_cfg['n_quantum_layers'],
                     entangling_layer=model_cfg['entangling_layer']
                 )
+            elif head_type == 'hamiltonian':
+                latent_dim = 2 ** n_qubits
+                H, W = grid_shape_for_amplitude(n_qubits)
+                if dataset_type == 'tcga':
+                    self.backbone = TCGABackbone(input_dim=768, latent_dim=latent_dim,
+                                                 preserve_spatial=True, out_grid=(H, W))
+                else:
+                    self.backbone = PCAMBackbone(latent_dim=latent_dim,
+                                                 filters=model_cfg['pcam_filters'],
+                                                 preserve_spatial=True, out_grid=(H, W))
+                variant = model_cfg.get('hamiltonian_variant', 'simple')  # 'simple' or 'paper'
+                if variant == 'simple':
+                    self.head = QuantumHeadHamiltonianSimple(
+                        n_qubits=n_qubits,
+                        num_classes=num_classes,
+                        n_layers=model_cfg['n_quantum_layers'],
+                        entangling_layer=model_cfg['entangling_layer'],
+                        include_zz=model_cfg.get('hamiltonian_include_zz', True),
+                        time=model_cfg.get('hamiltonian_time', 1.0)
+                    )
+                elif variant == 'paper':
+                    self.head = QuantumHeadHamiltonianPaper(
+                        n_qubits=n_qubits,
+                        num_classes=num_classes,
+                        n_layers=model_cfg['n_quantum_layers'],
+                        entangling_layer=model_cfg['entangling_layer'],
+                        include_xx=model_cfg.get('hamiltonian_include_xx', True),
+                        n_trotter_steps=model_cfg.get('hamiltonian_trotter_steps', 1),
+                        time=model_cfg.get('hamiltonian_time', 1.0)
+                    )
+                else:
+                    raise ValueError(f"Unknown hamiltonian_variant: {variant}")
             else:
                 raise ValueError(f"Unknown quantum_head_type: {head_type}")
         else:
             latent_dim = model_cfg['latent_dim']
             self.head = ClassicalHead(latent_dim, num_classes)
-            
+
     def forward(self, x: torch.Tensor) -> torch.Tensor:
         x = self.backbone(x)
```

Comment:
If I understand correctly you are using the classical encoder to go from the original picture to the latent space, and embed this latent space features with an Hamiltonian encoding in the quantum circuit. Am I correct ? That's not exactly what the paper is doing where they (iirc) directly embed the picture.